tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc

```diff
@@ -291,6 +285,47 @@ void UseInterval::intervalSubtract(UseInterval::Vector &a,
   }
 }
 
+/// Performs an interval intersection => A = A ^ B.
+void UseInterval::intervalIntersect(UseInterval::Vector &a,
+                                    const UseInterval::Vector &b) {
+  auto iterB = b.begin();
+  auto endB = b.end();
+  for (auto iterA = a.begin(); iterA != a.end() && iterB != endB;) {
+    // iterA is strictly before iterB => erase iterA.
+    if (*iterA < *iterB) {
+      iterA = a.erase(iterA);
+      // iterB is strictly before iterA => increment iterB.
+    } else if (*iterA > *iterB) {
+      ++iterB;
+      // iterA overlaps with the start of iterB. Set the start of iterA to the
+      // start of iterB and increment iterA.
+    } else if (iterA->start < iterB->start && iterA->end <= iterB->end) {
+      iterA->start = iterB->start;
+      ++iterA;
+      // iterA overlaps with the end of iterB.
+    } else if (iterA->end > iterB->end) {
+      // If iterA overlaps with the start of iterB, the start of iterA must be
+      // set to the start of iterB.
+      if (iterA->start < iterB->start)
+        iterA->start = iterB->start;
+      // The end of iterA must be set to the end of iterB. In addition, if iterB
+      // is not the end-pointer, the remaining interval of iterA must be added
+      // to vector A, because the next UseInterval of B can intersect with the
+      // remaining Interval.
+      size_t currentEndA = iterA->end;
+      iterA->end = iterB->end;
+      if (std::next(iterB) != endB)
+        iterA = a.insert(std::next(iterA),
+                         UseInterval(iterB->end + 1, currentEndA));
+      ++iterB;
+      // Both intervals are equal.
+    } else {
+      ++iterA;
+      ++iterB;
+    }
+  }
```

Comment:
If you get here and `iterA` is not at the end, then you have to drop the rest as it does not overlap.